aten/src/ATen/native/cpu/FlashAttentionKernel.cpp

```diff
@@ -0,0 +1,516 @@
+    return at::kFloat;
+  }
+}
+
+template <typename scalar_t>
+void _mha_mul_softmax_kernel(
+    float* a,
+    scalar_t* b,
+    float* dst,
+    float* max,
+    float* sum,
+    const float& scale,
+    const int& qsize,
+    const int& kvsize,
+    const int& headsize,
+    const int& idx) {
+  float tmp_max = 0.f, tmp_sum = 0.f, sum_old = 0.f, exp_tmp = 0.f;
+
+  for (int i = 0; i < qsize; ++i) {
+    sum_old = sum[i];
+
+    _mul_reduce_max_fusion_kernel(
+        a + i * kvsize, scale, kvsize, a + i * kvsize, tmp_max);
+
+    tmp_max = max[i] > tmp_max ? max[i] : tmp_max;
+
+    tmp_sum = tmp_max;
+    _exp_reduce_sum_fusion_kernel(
+        a + i * kvsize, kvsize, a + i * kvsize, tmp_sum);
+    exp_tmp = exp(max[i] - tmp_max);
+    sum[i] = tmp_sum + exp_tmp * sum[i];
+    max[i] = tmp_max;
+
+    _normalization_kernel<scalar_t>(
+        a + i * kvsize, sum[i], kvsize, b + i * kvsize);
+
+    if (idx) {
+      _mha_update_sum_max_kernel(
+        dst + i * headsize,
+        sum_old,
+        sum[i],
+        exp_tmp,
+        headsize,
+        dst + i * headsize);
+    }
+  }
+}
+
+template <typename scalar_t>
+void sd_mha_base_kernel(
+    scalar_t* output,
+    scalar_t* query,
+    scalar_t* key,
+    scalar_t* value,
+    const int64_t& qStride,
+    const int64_t& kStride,
+    const int64_t& vStride,
+    const int64_t& batchSize,
+    const int64_t& qSize,
+    const int64_t& kvSize,
+    const int64_t& num_head,
+    const int64_t& headSize,
+    const int64_t& hiddenSize,
+    const float& scale) {
+
+  int64_t qSplitSize = qSize;
+  for (size_t i = 0; i < qsplit_range.size(); ++i) {
+    if (qSize > qsplit_range[i]) {
+      qSplitSize = qsplit_size[i];
+      break;
+    }
+  }
+  int64_t kvSplitSize = kvSize >= kvsplit_size ? kvsplit_size : kvSize;
+
+  int64_t qSlice = (qSize - 1) / qSplitSize + 1;
+  int64_t qTail = (qSize - 1) % qSplitSize + 1;
+  int64_t kvSlice = (kvSize - 1) / kvSplitSize + 1;
+  int64_t kvTail = (kvSize - 1) % kvSplitSize + 1;
+
+  int64_t num_thread = at::get_num_threads();
+
+  at::Tensor qk = at::empty({num_thread, qSplitSize, kvSplitSize}, at::kFloat);
+  at::Tensor qk_norm = at::empty(
+      {num_thread, qSplitSize, kvSplitSize}, ConvertTypeToKType<scalar_t>());
+  at::Tensor qk_max = at::empty({num_thread, qSplitSize}, at::kFloat);
+  at::Tensor qk_sum = at::empty({num_thread, qSplitSize}, at::kFloat);
+  at::Tensor dst_fp32 =
+      at::empty({num_thread, qSplitSize, headSize}, at::kFloat);
+
+  at::parallel_for(0, batchSize * num_head * qSlice, 0, [&](int64_t begin, int64_t end) {
+    int64_t i = 0;
+    int64_t j = 0;
+    int64_t k = 0;
+    data_index_init(begin, i, batchSize, j, num_head, k, qSlice);
+
+    for (int64_t x = begin; x < end; x++) {
+      int qBlockSize = (k == qSlice - 1) ? qTail : qSplitSize;
+      int ompIdx = at::get_thread_num();
+      _init_mha_buffer_kernel(
+          qk_max.data_ptr<float>() + ompIdx * qSplitSize,
```

Comment:
`data_ptr` is a un-inlined function call, should be placed at inner loop since it is slow.